### PR TITLE
Strongly typed headers

### PR DESCRIPTION
**Pull request by Tratcher**
Consume the new strongly typed headers from HttpAbstractions.

Note MediaTypeWithQualityHeaderValue was merged into MediaTypeHeaderValue.

RE:
aspnet/HttpAbstractions#164 - New header types
#118 - Remove temp conneg types from MVC

### CODE REVIEW

**Comment by Tratcher** [comment]
@muratg @davidfowl @loudej @rynowak @yishaigalatzer

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs [hunk 1]
```diff
@@ -54,11 +55,12 @@ public override Task ExecuteResultAsync([NotNull] ActionContext context)
...
                 // basis for the actual filename, where possible.
-                var headerValue = ContentDispositionUtil.GetHeaderValue(FileDownloadName);
-                context.HttpContext.Response.Headers.Set("Content-Disposition", headerValue);
+                var cd = new ContentDispositionHeaderValue("attachment");
+                cd.SetHttpFileName(FileDownloadName);
+                context.HttpContext.Response.Headers.Set("Content-Disposition", cd.ToString());
             }
 
-            // We aren't flowing the cancellation token appropiately, see
+            // We aren't flowing the cancellation token appropriately, see
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: :100:

File: src/Microsoft.AspNet.Mvc.WebApiCompatShim/ContentNegotiator/DefaultContentNegotiator.cs [hunk 2]
```diff
@@ -319,32 +319,11 @@ public DefaultContentNegotiator(bool excludeMatchOnTypeOnly)
                 for (var i = 0; i < supportedMediaTypes.Count; i++)
                 {
                     var supportedMediaType = supportedMediaTypes[i];
-                    MediaTypeHeaderValueRange range;
-                    if (supportedMediaType != null &&
-                        acceptMediaTypeValue.Quality != FormattingUtilities.NoMatch &&
-                        supportedMediaType.IsSubsetOf(acceptMediaTypeValue, out range))
+                    MediaTypeFormatterMatchRanking ranking;
+                    if (supportedMediaType != null && acceptMediaTypeValue.Quality != FormattingUtilities.NoMatch &&
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.WebApiCompatShim/ContentNegotiator/DefaultContentNegotiator.cs]: Could you split the individual conditions across multiple lines like before?

File: src/Microsoft.AspNet.Mvc.WebApiCompatShim/ContentNegotiator/DefaultContentNegotiator.cs [hunk 3]
```diff
@@ -319,32 +319,11 @@ public DefaultContentNegotiator(bool excludeMatchOnTypeOnly)
...
-                                ranking = MediaTypeFormatterMatchRanking.MatchOnRequestAcceptHeaderLiteral;
-                                break;
-                        }
-
-                        return new MediaTypeFormatterMatch(
-                            formatter,
-                            supportedMediaType,
-                            acceptMediaTypeValue.Quality,
-                            ranking);
+                        return new MediaTypeFormatterMatch(formatter, supportedMediaType, acceptMediaTypeValue.Quality, ranking);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.WebApiCompatShim/ContentNegotiator/DefaultContentNegotiator.cs]: Split like before (since it exceeds 120 chars line limit we use).

File: src/Microsoft.AspNet.Mvc.WebApiCompatShim/ContentNegotiator/MediaTypeHeaderValueExtensions.cs [hunk 4]
```diff
@@ -49,40 +49,41 @@ public static bool IsSubsetOf(this MediaTypeHeaderValue mediaType1, MediaTypeHea
         /// media range.
         /// </param>
         /// <returns><c>true</c> if this is a subset of <paramref name="mediaType2"/>; false otherwise.</returns>
-        public static bool IsSubsetOf(
-            this MediaTypeHeaderValue mediaType1,
-            MediaTypeHeaderValue mediaType2,
-            out MediaTypeHeaderValueRange mediaType2Range)
+        public static bool IsSubsetOf(this MediaTypeHeaderValue mediaType1, MediaTypeHeaderValue mediaType2, out MediaTypeFormatterMatchRanking mediaType2Range)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.WebApiCompatShim/ContentNegotiator/MediaTypeHeaderValueExtensions.cs]: Line length.

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/CreatedAtActionResultTests.cs [hunk 5]
```diff
@@ -85,16 +84,21 @@ private static ActionContext GetActionContext(HttpContext httpContext)
                                     new ActionDescriptor());
         }
 
-        private static HttpContext GetHttpContext(HttpResponse response)
+        private static HttpContext GetHttpContext()
         {
             var httpContext = new Mock<HttpContext>();
-
+            var realContext = new DefaultHttpContext();
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/CreatedAtActionResultTests.cs]: Might be cleaner to simply return the `realContext` instance. All it would require is mocking the `Request.RequestServices` service provider.

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/JsonOutputFormatterTests.cs [hunk 6]
```diff
@@ -104,9 +104,9 @@ private static ActionContext GetActionContext(MediaTypeHeaderValue contentType)
         {
             var request = new Mock<HttpRequest>();
             var headers = new Mock<IHeaderDictionary>();
-            request.Setup(r => r.ContentType).Returns(contentType.RawValue);
+            request.Setup(r => r.ContentType).Returns(contentType.ToString());
             request.SetupGet(r => r.Headers).Returns(headers.Object);
-            request.SetupGet(f => f.AcceptCharset).Returns(contentType.Charset);
+            // TODO: request.SetupGet(f => f.AcceptCharset).Returns(contentType.Charset);
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/JsonOutputFormatterTests.cs]: What happened here?

**Comment by pranavkm** [comment]
:watch: 

@harshgMSFT, could you have a look at this?

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs [hunk 7]
```diff
@@ -32,25 +33,25 @@ public static IEnumerable<object[]> ContentTypes
                         };
 
                 // Empty accept header, should select based on contentTypes.
-                yield return new object[] { contentTypes, "", "application/json;charset=utf-8" };
+                yield return new object[] { contentTypes, "", "application/json; charset=utf-8" };
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs]: wondering why this line needed to change? do we expect a space between the parameters?

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs [hunk 8]
```diff
@@ -54,11 +55,12 @@ public override Task ExecuteResultAsync([NotNull] ActionContext context)
                 // detached and stored in a separate file. If the receiving MUA writes
                 // the entity to a file, the suggested filename should be used as a
                 // basis for the actual filename, where possible.
-                var headerValue = ContentDispositionUtil.GetHeaderValue(FileDownloadName);
-                context.HttpContext.Response.Headers.Set("Content-Disposition", headerValue);
+                var cd = new ContentDispositionHeaderValue("attachment");
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: should there be a default way/constructor to create one with attachment?

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs [hunk 9]
```diff
@@ -54,11 +55,12 @@ public override Task ExecuteResultAsync([NotNull] ActionContext context)
...
+                cd.SetHttpFileName(FileDownloadName);
+                context.HttpContext.Response.Headers.Set("Content-Disposition", cd.ToString());
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: Do we have no way to set the content disposition directly?

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs [hunk 10]
```diff
@@ -74,193 +76,5 @@ public override Task ExecuteResultAsync([NotNull] ActionContext context)
         /// A <see cref="Task"/> that will complete when the file has been written to the response.
         /// </returns>
         protected abstract Task WriteFileAsync(HttpResponse response, CancellationToken cancellation);
-
-        // This is a temporary implementation until we have the right abstractions in HttpAbstractions.
-        internal static class ContentDispositionUtil
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: yey

File: test/Microsoft.AspNet.Mvc.Core.Test/Filters/ProducesAttributeTests.cs [hunk 11]
```diff
@@ -40,22 +40,23 @@ public class ProducesAttributeTests
...
+            Assert.Equal(expectedMediaType.SubType, actualMediaType.SubType);
             Assert.Equal(expectedMediaType.Charset, actualMediaType.Charset);
-            Assert.Equal(expectedMediaType.MediaTypeRange, actualMediaType.MediaTypeRange);
+            Assert.Equal(expectedMediaType.AllTypes, actualMediaType.AllTypes);
+            Assert.Equal(expectedMediaType.AllSubTypes, actualMediaType.AllSubTypes);
             Assert.Equal(expectedMediaType.Parameters.Count, actualMediaType.Parameters.Count);
             foreach (var item in expectedMediaType.Parameters)
             {
-                Assert.Equal(item.Value, actualMediaType.Parameters[item.Key]);
+                // TODO: Assert.Equal(item.Value, actualMediaType.Parameters[item.Key]);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Filters/ProducesAttributeTests.cs]: Is there something tracking this?

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/JsonOutputFormatterTests.cs (hunk 6 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/JsonOutputFormatterTests.cs]: Missing parameter parsing???

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs (hunk 7 above)
Comment (Tratcher) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs]: Yes, the type now outputs spaces.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ConnegTests.cs [hunk 12]
```diff
@@ -105,8 +105,8 @@ public class ConnegTests
         }
 
         [Theory]
-        [InlineData("ContactInfoUsingV3Format", "text/vcard; charset=utf-8; version=v3.0", "BEGIN:VCARD#FN:John Williams#END:VCARD#")]
-        [InlineData("ContactInfoUsingV4Format", "text/vcard; charset=utf-8; version=v4.0", "BEGIN:VCARD#FN:John Williams#GENDER:M#END:VCARD#")]
+        [InlineData("ContactInfoUsingV3Format", "text/vcard; version=v3.0; charset=utf-8", "BEGIN:VCARD#FN:John Williams#END:VCARD#")]
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ConnegTests.cs]: why ? are we dependent on the order of parameters??

File: test/Microsoft.AspNet.Mvc.FunctionalTests/FileResultTests.cs [hunk 13]
```diff
@@ -58,7 +58,7 @@ public class FileResultTests
 
             var contentDisposition = response.Content.Headers.ContentDisposition.ToString();
             Assert.NotNull(contentDisposition);
-            Assert.Equal("attachment; filename=downloadName.txt", contentDisposition);
+            Assert.Equal("attachment; filename=downloadName.txt; filename*=UTF-8''downloadName.txt", contentDisposition);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/FileResultTests.cs]: seems unrelated?

**Comment by harshgMSFT** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs (hunk 9 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: response.ContentDisposition? No. response.GetTypedHeaders().ContentDisposition = cd would work, but it does the same thing as the above.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ConnegTests.cs (hunk 12 above)
Comment (Tratcher) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ConnegTests.cs]: ToString returns parameters in the order they were originally provided.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/FileResultTests.cs (hunk 13 above)
Comment (Tratcher) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/FileResultTests.cs]: ContentDispositionHeaderValue.SetHttpFileName intentionally sets both filename and filename* for broadest interop. The values will diverge if the file-name contains non-ascii.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/MediaTypeHeaderValueComparer.cs [hunk 14]
```diff
@@ -11,16 +11,16 @@ namespace Microsoft.AspNet.Mvc
     /// Implementation of <see cref="IComparer{T}"/> that can compare accept media type header fields
     /// based on their quality values (a.k.a q-values).
     /// </summary>
-    public class MediaTypeWithQualityHeaderValueComparer : IComparer<MediaTypeWithQualityHeaderValue>
+    public class MediaTypeHeaderValueComparer : IComparer<MediaTypeHeaderValue>
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/MediaTypeHeaderValueComparer.cs]: Should this go out of MVC?

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs (hunk 8 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: Proposals?
`new ContentDispositionHeaderValue()` is reserved for something else.
Maybe `ContentDispositionHeaderValue.CreateHttpAttachment(filename)`?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/MediaTypeHeaderValueComparer.cs (hunk 14 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/MediaTypeHeaderValueComparer.cs]: *shrug* I guess it would be useful for other people doing conneg.

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs (hunk 8 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: sure, or if possible `new ContentDispositionHeaderValue(string)` if it is not taken

File: src/Microsoft.AspNet.Mvc.Core/Formatters/MediaTypeHeaderValueComparer.cs (hunk 14 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/MediaTypeHeaderValueComparer.cs]: yeah it's not really important just weird that we expose it as a public class on MVC

**Comment by yishaigalatzer** [comment]
:shipit: when comments addressed and the rest of folks are happy

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs (hunk 8 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs]: `new ContentDispositionHeaderValue(string)` is taken as shown above. I recommend we postpone this and let customers try the existing version first.

**Comment by Eilon** [comment]
:shipit: from me too.
